Implement checked Shl/Shr on SIMD types

stygstra opened the pull request:
When overflow checking on `<<` and `>>` was added for integers, the `<<` and `>>` operations broke for SIMD types (`u32x4`, `i16x8`, etc.). This PR implements checked shifts on SIMD types.

Fixes #24258.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @huonw (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

stygstra commented:
The generated check is somewhat heavy: a checked shift on a `u8x16` inspects all 16 elements of the RHS. In general it generates `~2N` IR instructions for a checked shift on a `uMxN` or `iMxN`.

Is there a better way to check if a vector contains a nonzero element? Or is it not even worth worrying about? The check is only generated when `opt-level=0` or `debug-assertions` is enabled.

huonw commented:
I'm pretty sure there is a better way, but I'm very happy to just merge this as is having opened an issue and adding a `FIXME #<the issue number>`.

stygstra commented:
Figured it out. I didn't know about LLVM's `bitcast .. to` instruction.

huonw commented on src/librustc_trans/trans/expr.rs:
```diff
@@ -2477,9 +2475,44 @@ impl OverflowOpViaInputCheck {
+        TypeKind::Integer => ICmp(bcx, llvm::IntNE, value, C_null(llty), binop_debug_loc),
+        TypeKind::Vector => {
+            let width = llty.vector_length() as u64 * llty.element_type().int_width();
+            let int_value = BitCast(bcx, value, Type::ix(bcx.ccx(), width));
```
Neat trick; could you add a comment describing what's happening? ("check non-zeroness by treating the SIMD vector as one large integer and comparing that".)

stygstra replied:
(Added comment in latest commit)

huonw commented:
Looks great! Could you squash the commits down to one?

stygstra commented:
All set.

huonw commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 837c569 has been approved by `huonw`

<!-- @bors r=huonw 837c56961388b089343408dc8e1691a79ed1407f -->

bors commented:
:hourglass: Testing commit 837c569 with merge 38d1618...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5413)

stygstra commented:
The `core` feature was split apart in PR#26192. Rebased against master and replaced `#![feature(core)]` with `#![feature(core_simd)]` in the tests.

Running check-stage1 locally right now.

huonw commented:
@bors r+

bors commented:
:pushpin: Commit 875f50a has been approved by `huonw`

<!-- @bors r=huonw 875f50a8ee18297d7dc10e5fd7a200d902921f04 -->

bors commented:
:hourglass: Testing commit 875f50a with merge 84be3ef...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/398)

stygstra commented:
The failure looks similar to the example failure in PR #26382. Now that #26382 has been merged, try again?

alexcrichton commented:
@bors: retry

On Sat, Jun 20, 2015 at 9:34 AM, David Stygstra <notifications@github.com>
wrote:

> The failure looks similar to the example failure in PR #26382
> <https://github.com/rust-lang/rust/pull/26382>. Now that #26382
> <https://github.com/rust-lang/rust/pull/26382> has been merged, try again?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/26198#issuecomment-113788602>.
>

bors commented:
:hourglass: Testing commit 875f50a with merge cca2817...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/5436), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/5447), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/5398), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/5425), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5394), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/5468), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/5461), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5429), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/412), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/412), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/411), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/411)

stygstra commented:
Thanks!
